src/Documentation/Advanced-Concepts/Codegen.md

```diff
@@ -0,0 +1,16 @@
+---
+layout: page
+title: Code generation
+---
+
+# Code Generation
+
+**Efficient code generation is one of the pillars of Orleans Runtime**. Orleans Runtime make use of generated code in order to ensure proper serialization of types that are used across the cluster, and generating boilerplate that abstract from the Grain developer the need to understand how method shipping, exception propagation and other internal runtime concepts work. 
+
+There are two modes of `OrleansCodeGenerator`:
+
+1.	**Build-time Codegen** - in this mode, codegen will run every time your project is compiled. An `MSBuild` task is injected in your project's build pipeline, and the code is generated in the project's `Properties\orleans.codegen.cs`. To activate this mode, add the package `Microsoft.Orleans.OrleansCodeGenerator.Build` to your **Grain Interface** project and build it. You will notice that when adding this package no extra references are added to your project but you will see the `Properties\orleans.codegen.cs` was added. Also if you edit your `.csproj` file, you will see that an extra `MSBuild` target was added as well. The good thing of using this mode is that you can step-into the generated code while debugging since the file is physically on the disk however, your build time will be slower than usual if you have big projects. This is the default mode selected when you create a **Grain Interface** project by using the Visual Studio templates and it was the first codegen mode introduced since the release of Orleans. **Note**: To avoid the harness of keep this file under source control, we suggest you to to add it to your `.gitignore` or whatever ignore mechanism you have on your source control system, so it is not tracked by it, since this file will be re-generated every build.
```

Comment:
good thing of -> good thing about

physically on the disk however, -> physically on the disk. However, 

avoid the harness of keep -> avoid the hassles of keeping